### PR TITLE
(843-update-lookup-with-train-method

### PR DESCRIPTION
**Pull request by mborodii-prog**
# Update Lookup (with train method) 

## Overview
This pull request introduces support for an `action` parameter to the `write` method for training lookup models, allowing more granular control with `INSERT`, `UPDATE`, and `UPSERT` operations. The implementation includes robust error handling and validation for these actions, and comprehensive tests have been added to ensure correct behavior and coverage of edge cases.

### Feature: Action parameter for lookup model training

* Added `action` parameter to the `write` method in `wrangles/connectors/train.py`, supporting `INSERT`, `UPDATE`, and `UPSERT` actions for lookup models. This enables explicit control over whether to create, update, or upsert models. 
* Implemented logic for each action:
  - `UPSERT`: Updates existing models or creates new ones, handling duplicate keys and merging data.
  - `UPDATE`: Updates only existing records in a model, with validation for presence of keys.
  - `INSERT`: Creates a new model, with checks for duplicate model names and unique keys.
  - Added error handling for invalid combinations of parameters and unsupported actions.

## Usage Examples

# INSERT
```
  write:
    - train.lookup:
        name: My Lookup Wrangle
        action: INSERT
        variant: key
```

# UPDATE
```
  write:  
    - train.lookup:  
        model_id: test-model-id  
        action: UPDATE  
```

# UPSERT
```
  write:  
  - train.lookup:  
      name: {model_name} 
      action: UPSERT  
      variant: key  
```

# UPSERT (by default)
```
  write:  
  - train.lookup:  
      name: {model_name} 
      variant: key  
```

### Validation and error handling

* Comprehensive error messages and checks for duplicate model names, duplicate keys, missing models, and invalid action parameters, ensuring robust and user-friendly behavior.

### Test coverage

* Added extensive tests in `tests/connectors/test_train.py` for all new behaviors:
  - Successful insert and upsert operations.
  - Handling duplicate model names and duplicate keys.
  - Update operations for existing and non-existent models.
  - Validation of invalid actions and parameters.

### Documentation

* Updated schema documentation to include the new `action` parameter and its possible values, improving clarity for users.

### CODE REVIEW

**Comment by mborodii-prog** [comment]
@thomasstvr @ebhills Pls look into final version of PR:
- Implemented logic for each action:
UPSERT: Updates existing models or creates new ones, handling duplicate keys and merging data.
UPDATE: Updates only existing records in a model, with validation for presence of keys.
INSERT: Creates a new model, with checks for duplicate model names and unique keys.
OVERWRITE: Default logic (agreed with Thomas)


Mocked tests to avoid real model modification

File: wrangles/connectors/train.py [hunk 1]
```diff
@@ -242,33 +242,146 @@ def write(df: _pd.DataFrame, name: str = None, model_id: str = None, settings: d
...
+            if v == 'semantic':
+                v = 'embedding'
+            return v
+
+        def _set_variant(mid: str, var: str) -> str:
+            v = _normalize_variant(mid, var)
+            settings['variant'] = v
+            return v
+
+        if act == 'OVERWRITE':
```
Comment (thomasstvr) [inline wrangles/connectors/train.py]: Instead of checking for name or model_id below, name could be caught here since name will new models will be trained this way regardless of any action. 

```python
if act == 'OVERWRITE' or name:
```
Should work and would allow us to drop the checks further down.

File: wrangles/connectors/train.py [hunk 2]
```diff
@@ -242,33 +242,146 @@ def write(df: _pd.DataFrame, name: str = None, model_id: str = None, settings: d
...
+            else:
+                _set_variant(None, variant)
+                _train.lookup(new_data, name, None, settings)
+
+        elif act == 'UPDATE':
+            # Verify model exists
+            try:
+                metadata = _data.model(model_id)
+                if metadata.get('message') == 'error':
+                    raise ValueError(f"Lookup model '{model_id}' not found")
```
Comment (thomasstvr) [inline wrangles/connectors/train.py]: While I like the idea, this is actually incorrect. If the model_id exists, but the user does not have access then this error will still be triggered. Let's drop this error handling and let the "access denied" error pick it up.

File: wrangles/connectors/train.py [hunk 3]
```diff
@@ -242,33 +242,146 @@ def write(df: _pd.DataFrame, name: str = None, model_id: str = None, settings: d
...
+            _set_variant(model_id, variant)
+
+            _train.lookup(
+                _to_tight(df),
+                name,
+                model_id,
+                settings
+            )
+
+        elif act == 'UPSERT':
```
[the remark this answers is not in the document]
Comment (mborodii-prog) [reply wrangles/connectors/train.py]: now I did it next way:
let's say the model starts out with the table below: 
k1 v1
k2 v2
k3 v3

The user inserts the following data:
k1 v1_1
k2 v2_2
k4 v4


The ouput:
k1 v1_1
k2 v2_2
k3 v3
k4 v4

**Review by thomasstvr — CHANGES_REQUESTED** [review]

File: wrangles/connectors/train.py [hunk 4]
```diff
@@ -242,33 +242,211 @@ def write(df: _pd.DataFrame, name: str = None, model_id: str = None, settings: d
...
+                inserted = len(df)
+                _train.lookup(new_data, name, None, settings)
+                _logging.info(f"Lookup UPSERT (new): {inserted} rows inserted. Total rows: {inserted}.")
+
+        elif act == 'UPDATE':
+            # Verify model exists
+            try:
+                metadata = _data.model(model_id)
+            except Exception:
+                raise RuntimeError(f"Model with id {model_id} does not exist or access is demied.")
```
[the remark this answers is not in the document]
Comment (mborodii-prog) [reply wrangles/connectors/train.py]: @thomasstvr added all required changes
